src/model_theory/terms_and_formulas.lean

```diff
@@ -128,36 +129,76 @@ variables {L} {α : Type}
 section formula
 variable {n : ℕ}
 
-@[simps] instance : has_bot (L.bounded_formula α n) := ⟨bd_falsum⟩
+namespace bounded_formula
+
+instance : has_bot (L.bounded_formula α n) := ⟨bd_falsum⟩
 
 /-- The negation of a bounded formula is also a bounded formula. -/
-@[reducible] def bd_not (φ : L.bounded_formula α n) : L.bounded_formula α n := bd_imp φ ⊥
+def bd_not (φ : L.bounded_formula α n) : L.bounded_formula α n := bd_imp φ ⊥
+
+/-- Puts an `∃` quantifier on a bounded formula. -/
+def bd_exists (φ : L.bounded_formula α (n + 1)) : L.bounded_formula α n :=
+  bd_not (bd_all (bd_not φ))
+
+instance : has_top (L.bounded_formula α n) := ⟨bd_not ⊥⟩
 
-@[simps] instance : has_top (L.bounded_formula α n) := ⟨bd_not bd_falsum⟩
+instance : has_inf (L.bounded_formula α n) := ⟨λ f g, bd_not (bd_imp f (bd_not g))⟩
 
-@[simps] instance : has_inf (L.bounded_formula α n) := ⟨λ f g, bd_not (bd_imp f (bd_not g))⟩
+instance : has_sup (L.bounded_formula α n) := ⟨λ f g, bd_imp (bd_not f) g⟩
 
-@[simps] instance : has_sup (L.bounded_formula α n) := ⟨λ f g, bd_imp (bd_not f) g⟩
+/-- The biimplication between two bounded formulas. -/
+def bd_iff (φ ψ : L.bounded_formula α n) := φ.bd_imp ψ ⊓ ψ.bd_imp φ
+
+/-- A function to help relabel the variables in bounded formulas. -/
+def relabel_aux {n : ℕ} {α β : Type} (g : α → (β ⊕ fin n)) (k : ℕ) :
+  α ⊕ fin k → β ⊕ fin (n + k) :=
+(sum.map id fin_sum_fin_equiv) ∘ (equiv.sum_assoc _ _ _) ∘ (sum.map g id)
 
 /-- Relabels a bounded formula's variables along a particular function. -/
-@[simp] def bounded_formula.relabel {α β : Type} (g : α → β) :
-  ∀ {n : ℕ}, L.bounded_formula α n → L.bounded_formula β n
-| n bd_falsum := bd_falsum
-| n (bd_equal t₁ t₂) := bd_equal (t₁.relabel (sum.elim (sum.inl ∘ g) sum.inr))
-    (t₂.relabel (sum.elim (sum.inl ∘ g) sum.inr))
-| n (bd_rel R ts) := bd_rel R ((term.relabel (sum.elim (sum.inl ∘ g) sum.inr)) ∘ ts)
-| n (bd_imp f₁ f₂) := bd_imp f₁.relabel f₂.relabel
-| n (bd_all f) := bd_all f.relabel
+def relabel {α β : Type} (g : α → (β ⊕ fin n)) :
+  ∀ {k : ℕ}, L.bounded_formula α k → L.bounded_formula β (n + k)
+| k bd_falsum := bd_falsum
+| k (bd_equal t₁ t₂) := bd_equal (t₁.relabel (relabel_aux g k)) (t₂.relabel (relabel_aux g k))
+| k (bd_rel R ts) := bd_rel R (term.relabel (relabel_aux g k) ∘ ts)
+| k (bd_imp f₁ f₂) := bd_imp f₁.relabel f₂.relabel
+| k (bd_all f) := bd_all f.relabel
+
+/-- Places universal quantifiers on all extra variables of a bounded formula. -/
+def close_with_forall : L.bounded_formula α n → L.formula α :=
+nat.rec_on n id (λ n f φ, (f φ.bd_all))
```

Comment:
maybe use the equation compiler (pattern matching) so that you get some equations automatically? 
I think this was called `foralls` in Flypitch, not sure if that is a better name.

(same for next declaration)